File: backend/routers/clickhouse.py
"""ClickHouse credential storage endpoints.

Only CRUD for encrypted credentials — no query execution.
Queries run client-side from the browser via ClickHouse HTTP API.
"""

import logging

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from database import get_db
from models import ClickHouseConnection, User
from services.auth import get_current_user
from services.encryption import get_encryption

router = APIRouter(prefix="/clickhouse", tags=["clickhouse"])
logger = logging.getLogger(__name__)


class CreateConnectionRequest(BaseModel):
    name: str
    host: str
    port: int = 8443
    username: str
    password: str
    database: str | None = None
    secure: bool = True


class CreateConnectionResponse(BaseModel):
    id: str


class CredentialsResponse(BaseModel):
    host: str
    port: int
    username: str
    password: str
    database: str | None
    secure: bool


@router.post("/connections", response_model=CreateConnectionResponse)
async def create_connection(
    request: CreateConnectionRequest,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Create a ClickHouse connection with encrypted credentials."""
    encryption = get_encryption()
    password_encrypted, iv = encryption.encrypt(request.password)

    conn = ClickHouseConnection(
        user_id=user.id,
        name=request.name,
        host=request.host,
        port=request.port,
        username=request.username,
        password_encrypted=password_encrypted,
        encryption_iv=iv,
        database=request.database,
        secure=request.secure,
    )
    db.add(conn)
    await db.commit()

    logger.info(f"Created ClickHouse connection {conn.id} for user {user.id}")
    return CreateConnectionResponse(id=conn.id)


@router.get(
    "/connections/{connection_id}/credentials", response_model=CredentialsResponse
)
async def get_credentials(
    connection_id: str,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Fetch decrypted credentials for a connection. Password returned in memory only."""
    result = await db.execute(
        select(ClickHouseConnection).where(
            ClickHouseConnection.id == connection_id,
            ClickHouseConnection.user_id == user.id,
        )
    )
    conn = result.scalar_one_or_none()
    if not conn:
        raise HTTPException(status_code=404, detail="Connection not found")

    encryption = get_encryption()
    password = encryption.decrypt(conn.password_encrypted, conn.encryption_iv)

    return CredentialsResponse(
        host=conn.host,
        port=conn.port,
        username=conn.username,
        password=password,
        database=conn.database,
        secure=conn.secure,
    )


@router.delete("/connections/{connection_id}")
async def delete_connection(
    connection_id: str,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Delete a ClickHouse connection."""
    result = await db.execute(
        select(ClickHouseConnection).where(
            ClickHouseConnection.id == connection_id,
            ClickHouseConnection.user_id == user.id,
        )
    )
    conn = result.scalar_one_or_none()
    if not conn:
        raise HTTPException(status_code=404, detail="Connection not found")

    await db.delete(conn)
    await db.commit()

    logger.info(f"Deleted ClickHouse connection {connection_id} for user {user.id}")
    return {"status": "deleted"}

File: backend/src/routes/clickhouse.rs
//! ClickHouse credential storage endpoints.
//!
//! Only CRUD for encrypted credentials -- no query execution.
//! Queries run client-side from the browser via ClickHouse HTTP API.

use axum::extract::{Path, State};
use axum::http::StatusCode;
use axum::response::{IntoResponse, Response};
use axum::Json;
use serde::{Deserialize, Serialize};
use serde_json::json;
use uuid::Uuid;

use crate::auth::middleware::AuthUser;
use crate::error::error_response;
use crate::helpers::now_sqlite;
use crate::AppState;

// ---------------------------------------------------------------------------
// Row types (sqlx::FromRow)
// ---------------------------------------------------------------------------

#[derive(sqlx::FromRow)]
struct ClickHouseRow {
    #[allow(dead_code)]
    id: String,
    #[allow(dead_code)]
    user_id: String,
    host: String,
    port: i64,
    username: String,
    password_encrypted: Vec<u8>,
    encryption_iv: Vec<u8>,
    database: Option<String>,
    secure: bool,
}

// ---------------------------------------------------------------------------
// Request / Response models
// ---------------------------------------------------------------------------

#[derive(Deserialize)]
pub struct CreateConnectionRequest {
    name: String,
    host: String,
    #[serde(default = "default_port")]
    port: i64,
    username: String,
    password: String,
    database: Option<String>,
    #[serde(default = "default_secure")]
    secure: bool,
}

fn default_port() -> i64 {
    8443
}

fn default_secure() -> bool {
    true
}

#[derive(Serialize)]
struct CreateConnectionResponse {
    id: String,
}

#[derive(Serialize)]
struct CredentialsResponse {
    host: String,
    port: i64,
    username: String,
    password: String,
    database: Option<String>,
    secure: bool,
}

// ---------------------------------------------------------------------------
// Endpoints
// ---------------------------------------------------------------------------

/// POST /clickhouse/connections - create a ClickHouse connection with encrypted credentials.
pub async fn create_connection(
    State(state): State<AppState>,
    AuthUser(user): AuthUser,
    Json(body): Json<CreateConnectionRequest>,
) -> Result<impl IntoResponse, Response> {
    let enc = state.encryption.as_ref()
        .ok_or_else(|| error_response(StatusCode::INTERNAL_SERVER_ERROR, "Encryption not configured"))?;

    let (ciphertext, iv) = enc
        .encrypt(&body.password)
        .map_err(|e| error_response(StatusCode::INTERNAL_SERVER_ERROR, &format!("Encryption error: {e}")))?;

    let id = Uuid::new_v4().to_string();
    let now = now_sqlite();

    sqlx::query(
        "INSERT INTO clickhouse_connections
         (id, user_id, name, host, port, database, username, password_encrypted, encryption_iv, secure, created_at, updated_at)
         VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
    )
    .bind(&id)
    .bind(&user.id)
    .bind(&body.name)
    .bind(&body.host)
    .bind(body.port)
    .bind(&body.database)
    .bind(&body.username)
    .bind(&ciphertext)
    .bind(&iv)
    .bind(body.secure)
    .bind(&now)
    .bind(&now)
    .execute(&state.db)
    .await
    .map_err(|_| error_response(StatusCode::INTERNAL_SERVER_ERROR, "Database error"))?;

    Ok(Json(CreateConnectionResponse { id }))
}

/// GET /clickhouse/connections/{connection_id}/credentials - get decrypted credentials.
pub async fn get_credentials(
    State(state): State<AppState>,
    AuthUser(user): AuthUser,
    Path(connection_id): Path<String>,
) -> Result<impl IntoResponse, Response> {
    let row = sqlx::query_as::<_, ClickHouseRow>(
        "SELECT id, user_id, host, port, username, password_encrypted, encryption_iv, database, secure
         FROM clickhouse_connections WHERE id = ? AND user_id = ?",
    )
    .bind(&connection_id)
    .bind(&user.id)
    .fetch_optional(&state.db)
    .await
    .map_err(|_| error_response(StatusCode::INTERNAL_SERVER_ERROR, "Database error"))?
    .ok_or_else(|| error_response(StatusCode::NOT_FOUND, "Connection not found"))?;

    let enc = state.encryption.as_ref()
        .ok_or_else(|| error_response(StatusCode::INTERNAL_SERVER_ERROR, "Encryption not configured"))?;

    let password = enc
        .decrypt(&row.password_encrypted, &row.encryption_iv)
        .map_err(|e| error_response(StatusCode::INTERNAL_SERVER_ERROR, &format!("Decryption error: {e}")))?;

    Ok(Json(CredentialsResponse {
        host: row.host,
        port: row.port,
        username: row.username,
        password,
        database: row.database,
        secure: row.secure,
    }))
}

/// DELETE /clickhouse/connections/{connection_id} - delete a connection.
pub async fn delete_connection(
    State(state): State<AppState>,
    AuthUser(user): AuthUser,
    Path(connection_id): Path<String>,
) -> Result<impl IntoResponse, Response> {
    let result = sqlx::query(
        "DELETE FROM clickhouse_connections WHERE id = ? AND user_id = ?",
    )
    .bind(&connection_id)
    .bind(&user.id)
    .execute(&state.db)
    .await
    .map_err(|_| error_response(StatusCode::INTERNAL_SERVER_ERROR, "Database error"))?;

    if result.rows_affected() == 0 {
        return Err(error_response(StatusCode::NOT_FOUND, "Connection not found"));
    }

    Ok(Json(json!({"status": "deleted"})))
}
